### ISSUE TITLE
Question: Canary-v2 hallucination in streaming mode

### ISSUE THREAD

**Issue by azziko** [NVIDIA-NeMo/NeMo#15514]
Hello @andrusenkoau,

I did my own implementation of canary-v2 in simultaneous mode with alignatt using these PR’s combined https://github.com/NVIDIA-NeMo/NeMo/pull/15229 https://github.com/NVIDIA-NeMo/NeMo/pull/15449. They allow me to inspect xatt_scores and inject decoder input ids with partial transcription from the top level transcribe() method. 

I ran some extensive tests for ASR and AST. The ASR works great, compared to Whisper model, I got 3-4 percent of WER improvement on average(for chunk-size 1 sec and frame-threshold 5). 

However, for translation in the same scenario, canary hallucinates a lot(repeating words), and sometimes outputs whole sentences in the source language.

My question is, in your implementation with alignatt, have you noticed something similar? What did you do to avoid it? If not, what would be the most probable reason for that, so I could try and fix it?

Thank you beforehand

**Comment by hezeclark** [comment]
"AST (automatic speech translation) hallucination in streaming mode is a known failure mode for encoder-decoder models with attention-based alignment — the decoder can \"drift\" from the encoder states when processing partial inputs, especially when the source language's syntax diverges significantly from the target language.\n\n**Three root causes for streaming AST hallucination:**\n\n**1. Attention leakage across chunk boundaries**\n\nIn streaming mode, the model processes audio chunks independently. If chunk `N` ends mid-sentence in the source language but the decoder has already committed to a target-language sentence structure, chunk `N+1`'s decoder states will be misaligned. This causes the model to either:\n- Hallucinate filler words to complete the target sentence\n- Repeat the last few words from chunk `N` to \"re-anchor\" attention\n\n**Mitigation:** Use overlapping chunks with cross-fade to preserve context:\n\n```python\n# Instead of non-overlapping 1-second chunks:\n# [0.0s - 1.0s], [1.0s - 2.0s], [2.0s - 3.0s] → hard boundaries\n\n# Use 50% overlap with attention masking:\n# [0.0s - 1.5s], [0.5s - 2.0s], [1.0s - 2.5s] → soft boundaries\nchunk_size = 1.0\noverlap = 0.5\n\nfor start in np.arange(0, audio_duration, chunk_size - overlap):\n    end = start + chunk_size\n    chunk_audio = audio[int(start * sample_rate):int(end * sample_rate)]\n    \n    # Mask attention to the overlap region from the previous chunk\n    # to avoid double-counting encoder states\n    chunk_output = model.transcribe(\n        chunk_audio,\n        mask_overlap=True,\n        overlap_frames=int(overlap * sample_rate / hop_length)\n    )\n```\n\n**2. Decoder priming with source-language tokens**\n\nIf the model sees source-language text in the decoder input during training (e.g., for code-switching or mixed-language scenarios), it learns that outputting source-language tokens is valid. In streaming mode with partial inputs, the model may fall back to source-language tokens when it's uncertain about the target translation.\n\n**Check:** Does your decoder input during inference include any source-language tokens? If you're using `force_decoder_ids` or `decoder_start_token_id` that might contain language-specific tokens, try forcing the decoder to start with a target-language token:\n\n```python\n# Force decoder to start with target language token\nmodel.transcribe(\n    audio,\n    decoder_start_token_id=model.tokenizer.lang_token_to_id[\"<en>\"],  # Force English output\n    suppress_source_lang_tokens=True  # Prevent source language tokens in beam search\n)\n```\n\n**3. Frame threshold mismatch with translation latency**\n\nYou mention `frame-threshold 5` for ASR, which works well because transcription is nearly instantaneous (source audio → source text). But AST has inherent reordering — the target sentence structure may require waiting for more source context before committing to a translation. A frame threshold of 5 might force premature decoder commits.\n\n**Experiment:** Try increasing the frame threshold for AST to allow the model to buffer more encoder states before outputting target tokens:\n\n```python\n# ASR: low latency, low threshold\nasr_output = model.transcribe(audio, frame_threshold=5, task=\"asr\")\n\n# AST: higher latency, higher threshold to reduce hallucination\nast_output = model.transcribe(audio, frame_threshold=10, task=\"ast\")\n```\n\n**Debugging: inspect attention scores**\n\nSince you have access to `xatt_scores` via the custom alignatt PR, you can detect attention drift by checking if the attention distribution becomes uniform (model is \"lost\"):\n\n```python\nimport numpy as np\n\ndef attention_entropy(xatt_scores):\n    \"\"\"High entropy = uniform attention = model uncertainty\"\"\"\n    probs = np.exp(xatt_scores) / np.sum(np.exp(xatt_scores), axis=-1, keepdims=True)\n    entropy = -np.sum(probs * np.log(probs + 1e-10), axis=-1)\n    return entropy\n\nfor chunk_idx, xatt in enumerate(xatt_scores_per_chunk):\n    entropy = attention_entropy(xatt)\n    if entropy > threshold:  # e.g., threshold = 0.8 * max_entropy\n        print(f\"Chunk {chunk_idx}: High attention entropy, likely hallucinating\")\n```\n\nIf entropy spikes at the same time as hallucinations, it confirms attention drift.\n\n**Comparison to Whisper:** Whisper's decoder is trained with a \"no timestamps\" mode that encourages monotonic attention even in streaming scenarios. Canary-v2 might not have the same inductive bias, so attention can wander during AST. You could try adding a monotonic attention loss during fine-tuning if you're retraining the model.\n\n**Questions:**\n\n1. What's the source→target language pair? (e.g., Spanish→English vs Chinese→English have very different reordering complexity)\n2. Are you using beam search or greedy decoding? Beam search can amplify hallucinations by exploring low-probability paths.\n3. Does the hallucination happen at specific linguistic boundaries (e.g., end of clauses, after pronouns)?\n4. Have you tried the official NeMo streaming AST example (if one exists) to rule out implementation differences?\n\nIf increasing frame threshold + overlapping chunks doesn't resolve it, I'd suggest opening a separate issue with audio samples + attention visualizations so the NeMo team can debug the specific failure mode.\n"

**Comment by pzelasko** [comment]
CC @andrusenkoau

**Comment by andrusenkoau** [comment]
Hi @azziko,
It's cool that you achieved a good result with Canary ASR streaming using the AlignAtt policy!

Unfortunately, we didn't investigate the AST task much for the AlignAtt. But we need to remember that the Canary model was trained in fully offline mode, where the encoder had access to the entire input audio file. Using any streaming decoding techniques can trigger uncontrolled model behavior. It looks like the AST task is more sensitive to this.

I currently have no working hypotheses on how to fix this without retraining the model using the streaming encoder. Alternatively, you could try keeping all the left context (you can set 100s in the script) and see if this behavior persists for files up to 30 seconds long. It's possible that the loss of left context affects AST performance.

It would be great if you could figure this out!

**Comment by azziko** [comment]
Hi,

Thank you for your response. I found [Word-Boosting](https://docs.nvidia.com/nemo-framework/user-guide/latest/nemotoolkit/asr/asr_customization/word_boosting.html) feature and figured out the source words can be negative boosted before the start of the streaming. The approach is: negative boost some more frequent words in the source vocabulary(more frequent than presonal names .etc. This needs to be tested emperically) before the start of the streaming. 

I tested it manually as POC by unboosting the words from the golden transcript of the source lang part in my dataset, and it did work to some extent. I will run tests to measure to what extent, but seems promising. 

I will attach metrics here when I have them.

**Comment by azziko** [comment]
Increasing chunk-size and forced prefix size worked best. Still there is occasional language confusion, but I figured that's just the E2E native problem. Surprisingly, results outperform the offline translation, I wonder why:

High-latency
| Direction | chunk-size | frame-threshold | BLEU  | chrF  | LongYAAL(CU) | XCOMET-XL |
| --------- | ----- | ----- | ----- | ----- | ------------ | --------- |
| en -> de  | 3.0   | 16    | 31.73 | 60.83 | 3761.2       | 0.8776    |
| en -> it  | 3.0   | 16    | 43.56 | 68.32 | 3281.8       | 0.8227    |
| cs -> en  | 2.5   | 20    | 32.01 | 59.26 | 3641.3       | 0.8133    |

Low-latency:
| Direction | chunk-size | frame-threshold | BLEU  | chrF  | LongYAAL | XCOMET-XL |
| --------- | ----- | ----- | ----- | ----- | -------- | --------- |
| en -> de  | 1.5   | 4     | 20.70 | 52.60 | 1677.1   | 0.7744    |
| en -> it  | 1.0   | 8     | 34.79 | 62.21 | 1972.1   | 0.7618    |
| cs -> en  | 0.5   | 12    | 27.78 | 56.68 | 1996.9   | 0.7633    |

**Comment by andrusenkoau** [comment]
Hi @azziko, great results!

Did you use the Canary-v2 model with the AlignAtt decoding policy for the IWSLT challenge? 

> Surprisingly, results outperform the offline translation, I wonder why 

Do you mean high-latency decoding with AlignAtt versus full-audio decoding? What audio size did you use for decoding?

**Comment by azziko** [comment]
Hi @andrusenkoau, thanks!

Yes, I submitted pre-trained Canary-v2 to the IWSLT to see how it compares to the SoTA cascades. The main idea is that Canary is very light and can be run locally on a smartphone(which I’m working on the demo of).

The results would have probably been better with fine-tuning, but the pre-trained one seems to perform pretty well anyway.

> Do you mean high-latency decoding with AlignAtt versus full-audio decoding? What audio size did you use for decoding?

Yes, I meant the Canary’s chunking method for parallel processing of long audios(just calling transcribe on the full audio) against the high-latency AlignAtt. I read the paper report on Canary-v2 and I get it now why it is the case, so the question is resolved. Thanks

**Comment by andrusenkoau** [comment]
Thank you for sharing @azziko!

What cascade systems did you use for the streaming speech translation? It is quite interesting that the accuracy difference you observed compared to Canary streaming decoding.

Regarding the on-device application, it looks like this new work could be promising for this task: https://arxiv.org/abs/2602.24245. It is possible that the authors will make this model publicly available after additional development.

**Comment by azziko** [comment]
> What cascade systems did you use for the streaming speech translation? 

Our other main submission was Qwen3-1.7B ASR with alignatt into Gemma-E4B for translation. Canary we implement in E2E fashion, not a cascade. 

> It is quite interesting that the accuracy difference you observed compared to Canary streaming decoding.

I meant long-form inference described in section 6.4.1 of the report: https://arxiv.org/pdf/2509.14128. It’s effectively a Sliding Window approach described in https://arxiv.org/abs/2210.09754, which is good for SST, but is lacking compared to AlignAtt or another retranslation approach like Local Agreement, so I found the result natural. It’s a shame I overlooked the streaming decoding as another baseline.

> Regarding the on-device application, it looks like this new work could be promising for this task

Thank you! I will have a read
